src/serac/numerics/functional/tuple_tensor_dual_functions.hpp

```diff
@@ -0,0 +1,673 @@
+/// @p p.gradient = 1, then the @p x.gradient will be $dx/dp$.
+///
+/// The solver uses Newton's method, safeguarded by bisection. If the Newton update would take
+/// the next iterate out of the search interval, or the absolute value of the residual is not
+/// decreasing fast enough, bisection will be used to compute the next iterate. The bounds of the
+/// search interval are updated automatically to maintain a bracket around the root. If the sign
+/// of the residual is the same at both @p lower_bound and @p upper_bound, the solver aborts.
+template <typename function, typename... ParamTypes>
+auto solve_scalar_equation(function&& f, double x0, double tolerance, double lower_bound, double upper_bound,
+                           ParamTypes... params)
+{
+  double x, df_dx;
+  double fl = f(lower_bound, get_value(params)...);
+  double fh = f(upper_bound, get_value(params)...);
+
+  if (fl * fh > 0) {
+    SLIC_ERROR("solve_scalar_equation: root not bracketed by input bounds.");
+  }
+
+  unsigned int iterations = 0;
+  bool         converged  = false;
+
+  // handle corner cases where one of the brackets is the root
+  if (fl == 0) {
+    x         = lower_bound;
+    converged = true;
+  } else if (fh == 0) {
+    x         = upper_bound;
+    converged = true;
+  }
+
+  if (converged) {
+    df_dx = get_gradient(f(make_dual(x), get_value(params)...));
+
+  } else {
+    // orient search so that f(xl) < 0
+    double xl = lower_bound;
+    double xh = upper_bound;
+    if (fl > 0) {
+      xl = upper_bound;
+      xh = lower_bound;
+    }
+
+    // move initial guess if it is not between brackets
+    if (x0 < lower_bound || x0 > upper_bound) {
+      x0 = 0.5 * (lower_bound + upper_bound);
+    }
+    const unsigned int MAX_ITERATIONS = 25;
+    x                                 = x0;
+    double delta_x_old                = std::abs(upper_bound - lower_bound);
+    double delta_x                    = delta_x_old;
+    auto   R                          = f(make_dual(x), get_value(params)...);
+    auto   fval                       = get_value(R);
+    df_dx                             = get_gradient(R);
+
+    while (!converged) {
+      if (iterations == MAX_ITERATIONS) {
+        SLIC_WARNING("solve_scalar_equation failed to converge in allotted iterations.");
+        break;
+      }
+
+      // use bisection if Newton oversteps brackets or is not decreasing sufficiently
+      if ((x - xh) * df_dx - fval > 0 || (x - xl) * df_dx - fval < 0 ||
+          std::abs(2. * fval) > std::abs(delta_x_old * df_dx)) {
+        delta_x_old = delta_x;
+        delta_x     = 0.5 * (xh - xl);
+        x           = xl + delta_x;
+        converged   = (x == xl);
+      } else {  // use Newton step
+        delta_x_old = delta_x;
+        delta_x     = fval / df_dx;
+        auto temp   = x;
+        x -= delta_x;
+        converged = (x == temp);
+      }
+
+      // std::cout << "iter " << iterations << " x = " << x << std::endl;
```

Comment:
dead code

Reply:
How fine-grained are the logging tools? Output of this kind can be useful for debugging. You wouldn't want to turn it on in a finite element simulation (even in debug) where this is operative at quadrature points, because you'll be flooded with data. On the other hand, if you're debugging a material model, you might want this information.

I'll remove it for now, we can discuss this when we we need it.